rust-utils/src/launcher_lair_client.rs

```diff
@@ -0,0 +1,331 @@
+        }
+
+        Ok(imported_pubkey_b64.to_string())
+    }
+}
+
+#[napi(js_name = "LauncherLairClient")]
+pub struct JsLauncherLairClient {
+    launcher_lair_client: Option<LauncherLairClient>,
+}
+
+#[napi]
+impl JsLauncherLairClient {
+    #[napi(constructor)]
+    pub fn new() -> Self {
+        Self {
+            launcher_lair_client: None,
+        }
+    }
+
+    #[napi]
+    pub async fn connect(connection_url: String, passphrase: String) -> Self {
+        let launcher_lair_client = LauncherLairClient::new(connection_url, passphrase).await;
+
+        JsLauncherLairClient {
+            launcher_lair_client: Some(launcher_lair_client),
+        }
+    }
+
+    #[napi]
+    pub async fn sign_zome_call(
+        &self,
+        zome_call_unsigned_js: ZomeCallUnsignedNapi,
+    ) -> Result<ZomeCallNapi> {
+        self.launcher_lair_client
+            .as_ref()
+            .unwrap()
+            .sign_zome_call(zome_call_unsigned_js)
+            .await
+    }
+
+    #[napi]
+    pub async fn derive_and_import_seed_from_json_file(
+        &self,
+        path: String,
+    ) -> Result<String> {
+        self.launcher_lair_client
+            .as_ref()
+            .unwrap()
+            .derive_and_import_seed_from_json_file(path)
+            .await
+    }
+}
+
+/// Unlocks the device bundle seed with the given passphrase, then derives
+/// a sub SeedBundle by the given index and returns the corresponding derived KeyPair
+pub async fn derive_seed_from_device_bundle(
+    device_bundle: &String,
+    passphrase: SecretString,
+    index: u32,
+) -> napi::Result<Keypair> {
+    let cipher = base64::decode(device_bundle).map_err(|e| {
+        napi::Error::from_reason(format!("Failed to decode device bundle: {:?}", e))
+    })?;
+    match UnlockedSeedBundle::from_locked(&cipher)
+        .await
+        .map_err(|e| {
+            napi::Error::from_reason(format!("Failed to create UnlockedSeedBundle: {:?}", e))
+        })?
+        .remove(0)
+    {
+        LockedSeedCipher::PwHash(cipher) => {
+            let passphrase_bufread = BufRead::from(passphrase.expose_secret().as_bytes());
+            let seed = cipher.unlock(passphrase_bufread).await.map_err(|e| {
+                napi::Error::from_reason(format!(
+                    "Failed to unlock cipher with the given passphrase: {:?}",
+                    e
+                ))
+            })?;
+            let derived_seed = seed.derive(index).await.unwrap();
+
+            Ok(Keypair {
+                public: PublicKey::from_bytes(&*derived_seed.get_sign_pub_key().read_lock())
+                    .map_err(|e| {
+                        napi::Error::from_reason(format!("Failed to get public key: {:?}", e))
+                    })?,
+                secret: SecretKey::from_bytes(&*derived_seed.get_seed().read_lock()).map_err(
```

Comment:
Here you are taking the secret out of protected memory, and placing it into unprotected memory. The protected memory struct can be passed directly into `crypto_box_xsalsa_by_pub_key` so it never has to be unprotected.

Reply:
But, I guess if you need it for the dalek library anyway, you can keep this as-is. Another option, though, would be to use sodoken directly instead of dalek.

Reply:
Alright, thanks. I think I can do that, it's only the public key that I would ideally want to have in plain format to be able to compare with what's expected before importing it.

Reply:
It looks like LairClient currently can't accept a sodoken buffer directly, so I guess you'll need to leave this as-is unless/until we update that api to be more flexible.